### PR TITLE
startChat and endChat Broadcast event support

### PR DESCRIPTION
**Pull request by abhinab22mohanty**
### Added the following:

1. Added broadcasting event support for startChat and endChat.
2. Added custom Component Override flexibility for complete preChat Survey Pane.
3. Added custom widget story to external storybook.
4. Added Telemetry event for PostChat Loading Pane Load

Storybook Custom widget screenshots:
image
image
image
image
image
image

### CODE REVIEW

File: chat-widget/src/common/telemetry/TelemetryConstants.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ export enum LogLevel {
 // Events used in certain functionalities that are not being logged
 export enum BroadcastEvent {
     LoadPostChatSurvey = "LoadPostChatSurvey",
-    EndChat = "EndChat",
+    EndChat = "ChatEnded",
```
Comment (charliewang95) [inline chat-widget/src/common/telemetry/TelemetryConstants.ts:26]: Telemetry doc change for this?

File: chat-widget/src/common/telemetry/TelemetryConstants.ts [hunk 2]
```diff
@@ -82,7 +82,9 @@ export enum TelemetryEvent {
     PrechatSurveyLoaded = "PrechatSurveyLoaded",
     PrechatSubmitted = "PrechatSubmitted",
     StartChatSDKCall = "StartChatCall",
+    StartChatEventRecevied = "StartChatEventReceived",
```
Comment (charliewang95) [inline chat-widget/src/common/telemetry/TelemetryConstants.ts:85]: doc update

File: chat-widget/src/components/headerstateful/HeaderStateful.tsx [hunk 3]
```diff
@@ -36,10 +36,7 @@ export const HeaderStateful = (props: IHeaderStatefulParams) => {
             TelemetryHelper.logActionEvent(LogLevel.INFO, { Event: TelemetryEvent.HeaderCloseButtonClicked, Description: "Header Close button clicked." });
             if (conversationState.current === ConversationState.Active) {
                 dispatch({ type: LiveChatWidgetActionType.SET_SHOW_CONFIRMATION, payload: true });
-            } else if (conversationState.current === ConversationState.Postchat) {
```
Comment (charliewang95) [inline chat-widget/src/components/headerstateful/HeaderStateful.tsx:39]: Why this change?

File: chat-widget/src/components/livechatwidget/common/endChat.ts [hunk 4]
```diff
@@ -9,9 +9,35 @@ import { LiveChatWidgetActionType } from "../../../contexts/common/LiveChatWidge
...
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-export const endChat = async (props: ILiveChatWidgetProps, chatSDK: any, setAdapter: any, setWebChatStyles: any, dispatch: Dispatch<ILiveChatWidgetAction>, adapter: any, skipEndChatSDK?: boolean, skipCloseChat?: boolean) => {
+const prepareEndChat = async (props: ILiveChatWidgetProps, chatSDK: any, setAdapter: any, setWebChatStyles: any, dispatch: Dispatch<ILiveChatWidgetAction>, adapter: any, state: ILiveChatWidgetContext) => {
+    const isPostChatEnabled = state.domainStates.liveChatConfig?.LiveWSAndLiveChatEngJoin?.msdyn_postconversationsurveyenable;
+    const postChatSurveyMode = state.domainStates.liveChatConfig?.LiveWSAndLiveChatEngJoin?.msdyn_postconversationsurveymode;
+    const conversationDetails = await chatSDK.getConversationDetails();
+    if (isPostChatEnabled === "true" && conversationDetails?.canRenderPostChat === Constants.truePascal) {
+        const skipEndChatSDK = false;
+        const skipCloseChat = true;
+        await endChat(props, chatSDK, setAdapter, setWebChatStyles, dispatch, adapter, skipEndChatSDK, skipCloseChat);
```
Comment (charliewang95) [inline chat-widget/src/components/livechatwidget/common/endChat.ts:25]: Have you tested linked mode post chat survey after you move this line before checking for post chat mode?

File: chat-widget/src/components/livechatwidget/livechatwidgetstateful/LiveChatWidgetStateful.tsx [hunk 5]
```diff
@@ -140,6 +141,36 @@ export const LiveChatWidgetStateful = (props: ILiveChatWidgetProps) => {
                 });
             }
         });
+
+        // start chat from SDK Event
+        BroadcastService.getMessageByEventName("StartChat").subscribe((msg: ICustomEvent) => {
```
Comment (charliewang95) [inline chat-widget/src/components/livechatwidget/livechatwidgetstateful/LiveChatWidgetStateful.tsx:146]: please help suppress or fix all tslint warnings

File: chat-widget/src/components/livechatwidget/livechatwidgetstateful/LiveChatWidgetStateful.tsx [hunk 6]
```diff
@@ -140,6 +141,36 @@ export const LiveChatWidgetStateful = (props: ILiveChatWidgetProps) => {
...
+            TelemetryHelper.logActionEvent(LogLevel.INFO, {
+                Event: TelemetryEvent.StartChatEventRecevied,
+                Description: "Start chat event received."
+            });
+            if (state.appStates.isMinimized) {
+                dispatch({ type: LiveChatWidgetActionType.SET_MINIMIZED, payload: false });
+            } else {
+                prepareStartChat(props, chatSDK, state, dispatch, setAdapter);
+            }
+
```
Comment (charliewang95) [inline chat-widget/src/components/livechatwidget/livechatwidgetstateful/LiveChatWidgetStateful.tsx:156]: remove empty line

File: chat-widget/src/components/livechatwidget/interfaces/ILiveChatWidgetComponentOverrides.ts [hunk 7]
```diff
@@ -10,6 +10,7 @@ export interface ILiveChatWidgetComponentOverrides {
     outOfOfficeHoursPane?: ReactNode | string;
     postChatLoadingPane?: ReactNode | string;
     postChatSurveyPane?: ReactNode | string;
+    preChatSurveyPane?: ReactNode | string;
```
Comment (charliewang95) [inline chat-widget/src/components/livechatwidget/interfaces/ILiveChatWidgetComponentOverrides.ts:13]: Why add this now?

File: chat-widget/stories/OmniChannelLiveChatWidget.stories.tsx [hunk 8]
```diff
@@ -113,4 +113,242 @@ const liveChatWidgetDefaultWithSurveyProps: ILiveChatWidgetProps = {
     telemetryConfig: undefined
 };
 
-DefaultWithSurvey.args = liveChatWidgetDefaultWithSurveyProps;
\ No newline at end of file
+DefaultWithSurvey.args = liveChatWidgetDefaultWithSurveyProps;
```
Comment (charliewang95) [inline chat-widget/stories/OmniChannelLiveChatWidget.stories.tsx:116]: Please share some screenshots on different components for this custom widget (please link in description), thanks!

File: chat-widget/yarn.lock [hunk 9]
```diff
@@ -5994,6 +5994,13 @@ create-hmac@^1.1.0, create-hmac@^1.1.4, create-hmac@^1.1.7:
     safe-buffer "^5.0.1"
     sha.js "^2.4.8"
 
+cross-fetch@3.1.4:
```
Comment (charliewang95) [inline chat-widget/yarn.lock:5997]: Why is yarn.lock modified? Can we revert it?

File: chat-widget/src/components/headerstateful/HeaderStateful.tsx (hunk 3 above)
Comment (abhinab22mohanty) [reply chat-widget/src/components/headerstateful/HeaderStateful.tsx:39]: This changed is to remove redundant logic now that I have moved the PostChat check into EndChat

File: chat-widget/src/components/livechatwidget/common/endChat.ts (hunk 4 above)
Comment (abhinab22mohanty) [reply chat-widget/src/components/livechatwidget/common/endChat.ts:25]: Yes tested both scenarios

File: chat-widget/src/components/livechatwidget/interfaces/ILiveChatWidgetComponentOverrides.ts (hunk 7 above)
Comment (abhinab22mohanty) [reply chat-widget/src/components/livechatwidget/interfaces/ILiveChatWidgetComponentOverrides.ts:13]: This is to add support for one of our customers that might need the custom Override for PreChat as well.

File: chat-widget/yarn.lock (hunk 9 above)
Comment (abhinab22mohanty) [reply chat-widget/yarn.lock:5997]: Seems like with the existing package.json file running a yarn install is updating the lockfile.
This is to keep the lockfile updated

File: chat-widget/stories/OmniChannelLiveChatWidget.stories.tsx (hunk 8 above)
Comment (abhinab22mohanty) [reply chat-widget/stories/OmniChannelLiveChatWidget.stories.tsx:116]: Updated

**Review by charliewang95 — APPROVED** [review]
